### PR TITLE
Compile and cache configFile eventhandlers at preStart

### PR DESCRIPTION
**Pull request by dedmen**
They won't change after preStart as the config cannot be edited. So no reason to recompile them each preInit.

This is a mission with no eventhandlers in description or campaign config. Meaning 100% of that compileEventHandlers is just configFile. 670ms wasted each preInit for something that can be done once.

### CODE REVIEW

**Comment by dedmen** [comment]
Before:


After:

tracy_2019-02-08_11-42-14

File: addons/xeh/fnc_preStart.sqf [hunk 1]
```diff
@@ -65,4 +65,7 @@ with uiNamespace do {
             diag_log text format ["[XEH]: %1 does not support Extended Event Handlers! Addon: %2", _classname, _addon];
         };
     } forEach (true call CBA_fnc_supportMonitor);
+
+    //compile and cache configFile eventhandlers as they won't change from here on
+    GVAR(configFileEventHandlers) = compileFinal str (configFile call CBA_fnc_compileEventHandlers);
```
Comment (PabstMirror) [inline addons/xeh/fnc_preStart.sqf]: This func does several `isServer`/`isDedicated` checks to selectivly compile
`serverInit` xeh
https://github.com/CBATeam/CBA_A3/blob/master/addons/xeh/fnc_compileEventHandlers.sqf#L48-L57
At preStart everyone is a server

File: addons/xeh/fnc_compileEventHandlers.sqf [hunk 2]
```diff
@@ -151,21 +147,17 @@ private _resultNames = [];
                 };
 
                 // client only events
-                if (!isDedicated) then {
-                    _entry = _x >> format ["client%1", _entryName];
+                _entry = _x >> "clientInit";
```
Comment (commy2) [inline addons/xeh/fnc_compileEventHandlers.sqf]: format ["client%1", _entryName];

File: addons/xeh/fnc_compileEventHandlers.sqf [hunk 3]
```diff
@@ -151,21 +147,17 @@ private _resultNames = [];
...
-                        _eventFunc = _eventFunc + getText _entry + ";";
-                    };
+                if (isText _entry) then {
+                    _eventFunc = _eventFunc + "if (!isDedicated) then {" + getText _entry + "};";
                 };
 
                 // server only events
-                if (isServer) then {
-                    _entry = _x >> format ["server%1", _entryName];
+                _entry = _x >> "serverInit";
```
Comment (commy2) [inline addons/xeh/fnc_compileEventHandlers.sqf]: format ["server%1", _entryName];

File: addons/xeh/fnc_compileEventHandlers.sqf (hunk 2 above)
Comment (dedmen) [reply addons/xeh/fnc_compileEventHandlers.sqf]: I don't even know how I managed to do that

Comment (dedmen) [reply addons/xeh/fnc_compileEventHandlers.sqf]: Ah I just copied the stuff from the non-object eventhandlers. They don't have that stuff apparently

**Comment by PabstMirror** [comment]
rough idea: https://github.com/CBATeam/CBA_A3/commit/06ad425c1815233d3446f9e57185b150a21f81ca
compile server/client xeh seperately and selectivly add at mission time
I skipped pre/postInit events for now because that won't really effect performance

**Comment by commy2** [comment]
Anyone update this PR, or should I make my own with both ideas?
This is the last one we need for 3.10 I'd say.

**Comment by dedmen** [comment]
Can't you just push into this one? feel free to. I won't have time till monday

**Comment by commy2** [comment]
How?

**Comment by PabstMirror** [comment]
`git push https://github.com/dedmen/CBA_A3.git YourLocalBranchName:cacheConfigCompileEH`

**Comment by PabstMirror** [comment]
pushed changes for the init events as well

`if (!(_funcAll isEqualTo {})) then {` is pretty much the same as `[] call {}`, 
so we could remove those checks (for the init, not classEvents) if that looks better

preProcessing is disabled with the recompile bit or `filePatching` to prevent problems with developers
also added some debugging for funcs that can be precompiled
```
XEH: Could not recompile [postInit-cba_versioning]: if !(isServer) then {call compile preprocessFileLineNumbers '\x\cba\addons\versioning\XEH_postInitClient.sqf'}
```

**Comment by commy2** [comment]
```
$ git push https://github.com/dedmen/CBA_A3.git dedmen-cacheConfigCompileEH
Counting objects: 81, done.
Delta compression using up to 6 threads.
Compressing objects: 100% (81/81), done.
Writing objects: 100% (81/81), 12.79 KiB | 2.56 MiB/s, done.
Total 81 (delta 65), reused 0 (delta 0)
remote: Resolving deltas: 100% (65/65), completed with 37 local objects.
To https://github.com/dedmen/CBA_A3.git
 ! [remote rejected]   dedmen-cacheConfigCompileEH -> dedmen-cacheConfigCompileEH (permission denied)
error: failed to push some refs to 'https://github.com/dedmen/CBA_A3.git'
```

**Comment by commy2** [comment]
That didn't work well. Now the diff shows commits between when this branch was split from master and not the current master.

File: addons/xeh/fnc_initDisplay.sqf [hunk 4]
```diff
@@ -15,5 +15,5 @@ if !(_event isEqualTo "") then {
             private ["_event", "_className"]; // prevent these variables from being overwritten
             _args call compile getText _x;
         } forEach configProperties [_x >> XEH_FORMAT_CONFIG_NAME(_event) >> _className, "isText _x"];
-    } forEach XEH_MAIN_CONFIGS;
```
Comment (dedmen) [inline addons/xeh/fnc_initDisplay.sqf]: why?

Comment (commy2) [reply addons/xeh/fnc_initDisplay.sqf]: Macro seemed pointless. Bugged me for a while and now is an opportunity to muck it out.

**Comment by commy2** [comment]
By sorting the events in global/client/server on preStart in 3 different functions, we change the order in which these functions are executed. It used to be in order they appear in config regardless whether it was a global or server event. The only thing that mattered was load order.
Now all global ones are executed first, then all client ones, then all server ones.

**Comment by PabstMirror** [comment]
```
class Extended_PostInit_EventHandlers {
    class test1 {
        init = "diag_log 'test1 - init'";
        serverInit = "diag_log 'test1 - serverInit'";
    };    
    class test2 {
        init = "diag_log 'test2 - init'";
        serverInit = "diag_log 'test2 - serverInit'";
    };
    class test3 {
        init = "diag_log 'test3 - init'";
        serverInit = "diag_log 'test3 - serverInit'";
    };
    test4 = "diag_log 'test4 - init'";
};
```
"test1 - init"
"test1 - serverInit"
"test2 - init"
"test2 - serverInit"
"test3 - init"
"test3 - serverInit"
"test4 - init"

**Pull request merged**